### PR TITLE
Add guidelines for running the tutorial locally

### PR DESCRIPTION
**Pull request by a-roberts**
For https://github.com/knative/build-pipeline/issues/410, mentions which things we need to change and adds a section which includes running on Docker for Desktop.

Would be useful to have somebody else follow same instructions before we look at getting this merged please.

### CODE REVIEW

File: docs/tutorial.md [hunk 1]
```diff
@@ -291,7 +293,7 @@ spec:
           name: skaffold-image-leeroy-web
   results:
     type: gcs
-    url: gcs://somebucket/results/logs
+    url: gcs://somebucket/results/logs #configure: remove results entirely if you're happy to use stdout
```
Comment (bobcatfish) [inline docs/tutorial.md]: note to whoever implements #386 that we might want to remove the `results` config from our docs since it currently does nothing :thinking:

File: docs/tutorial.md [hunk 2]
```diff
@@ -556,3 +558,29 @@ status:
...
+
+
+## Local development
+
+### Known good configuration
+
+Knative (as of version 0.3) is known to work with:
+- [Docker for Desktop](https://www.docker.com/products/docker-desktop): a version that uses Kubernetes 1.11 or higher. At the time of this document, this requires the *edge* version of Docker to be installed
+- The following [prerequisites](https://github.com/knative/build-pipeline/blob/master/DEVELOPMENT.md#requirements)
+- Setting `host.docker.local:5000` as an insecure registry with Docker for Desktop
```
Comment (bobcatfish) [inline docs/tutorial.md]: do you happen to have a link to docs on how to set this as an insecure registry?

File: docs/tutorial.md [hunk 3]
```diff
@@ -556,3 +558,29 @@ status:
...
+- A local Docker registry: this can be run with
+
+`docker run -d -p 5000:5000 --name registry-srv -e REGISTRY_STORAGE_DELETE_ENABLED=true registry:2`
+
+### Images
+- Any hardcoded image locations should be replaced with `host.docker.internal:5000/myregistry/<image name>` equivalents: look for `PipelineResource` definitions which define an `image` specification
```
Comment (bobcatfish) [inline docs/tutorial.md]: i might reword this as something like "Any `PipelineResource` definitions of `image` type should be updated to use the local registry by setting the url to  `host.docker.internal:5000/myregistry/<image name>` "

this is a total nitpick but I wanted to point it out b/c flexibility of handling different registries with minimal Pipeline changes is a goal :D

File: docs/tutorial.md (hunk 2 above)
Comment (a-roberts) [reply docs/tutorial.md]: There's a lil whale we can click under Docker for Desktop -> Daemon, the documentation [here](https://docs.docker.com/registry/insecure/) isn't the most concise or I'd gladly refer to this. I'll mention it's under the preferences - I also need to shout out that running knative and istio requires *lots* of CPU and memory!

File: docs/tutorial.md (hunk 3 above)
Comment (a-roberts) [reply docs/tutorial.md]: Yeah you're exactly right, adding, ty

**Comment by a-roberts** [comment]
@bobcatfish I'd love for other devs to give this a go so please do advertise it in the Slack, on the case for the conflict and your feedback now and yes, documentation for running pipelines in general is a good shout I think. So would that be under https://github.com/knative/docs and into the main readme? Or are you thinking the main knative docs repo?

**Comment by bobcatfish** [comment]
> So would that be under https://github.com/knative/docs and into the main readme? Or are you thinking the main knative build-pipeline repo's readme?

I think for now we want to keep the docs in this repo - if we DO start having them in knative/docs, we'll probably do some kind of regular copying/mirroring process (as part of releasing)

> @bobcatfish I'd love for other devs to give this a go so please do advertise it in the Slack,

kk I will put out a call!

File: docs/tutorial.md [hunk 4]
```diff
@@ -560,3 +562,34 @@ status:
...
+
+- Optionally, a Docker registry viewer so we can check our pushed images are present:
+
+`docker run -it -p 8080:8080 --name registry-web --link registry-srv -e REGISTRY_URL=http://registry-srv:5000/v2 -e REGISTRY_NAME=localhost:5000 hyper/docker-registry-web`
+
+### Images
+- Any PipelineResource definitions of image type should be updated to use the local registry by setting the url to `host.docker.internal:5000/myregistry/<image name>` equivalents
+- The `KO_DOCKER_REPO` variable should be set to `host.docker.internal:5000/myregistry` before using `ko`
+- You are able to push to `host.docker.internal:5000/myregistry/<image name>` but your applications (e.g any deployment definitions) should reference `localhost:5000/myregistry/<image name>`
+
```
Comment (objectiveous) [inline docs/tutorial.md]: This PR is really helpful, thank you! I did run into a possible issue: 

I think KO_DOCKER_REPO on the host machine (macos) needs to be set to `localhost:5000/myregistry` NOT `host.docker.internal:5000/myregistry`. As the doc reads now, the build-pipeline installation call to `ko apply -f config/` will fail.

Comment (a-roberts) [reply docs/tutorial.md]: @objectiveous, you are exactly right, let me add that immediately!

File: docs/tutorial.md [hunk 5]
```diff
@@ -561,6 +563,36 @@ status:
...
+### Images
+- Any PipelineResource definitions of image type should be updated to use the local registry by setting the url to `host.docker.internal:5000/myregistry/<image name>` equivalents
+- The `KO_DOCKER_REPO` variable should be set to `localhost:5000/myregistry` before using `ko`
+- You are able to push to `host.docker.internal:5000/myregistry/<image name>` but your applications (e.g any deployment definitions) should reference `localhost:5000/myregistry/<image name>`
+
+### Logging
+- Logs can remain in-memory only as opposed to sent to a service such as [Stackdriver](https://cloud.google.com/logging/). Achieve this by modifying or deleting entirely (to just use stdout) a PipelineRun or TaskRun's `results` specification.
+
+Elasticsearch can be deployed locally as a means to view logs "after the fact": an example is provided at https://github.com/mgreau/knative-elastic-tutorials.
```
Comment (mgreau) [inline docs/tutorial.md]: Thank you very much for adding a link to these tutorials here ❤️ 
I have added a [kaniko tutorial](https://github.com/mgreau/knative-elastic-tutorials/blob/master/pipeline-elastic/README.adoc#kaniko-hello-world---dockerhub) and you can see [in this demo](https://github.com/mgreau/knative-elastic-tutorials/blob/master/demo/knative-kaniko-elastic-demo.mp4), it's possbile to stream the logs in real-time.

**Comment by bobcatfish** [comment]
wut

/test pull-knative-build-pipeline-integration-tests

**Comment by bobcatfish** [comment]
Thanks so much for taking a look at this @objectiveous and @mgreau !

@a-roberts let me know when you want another lgtm

**Comment by a-roberts** [comment]
@bobcatfish I'm ready for another, thanks for taking a look all 🎉

**Comment by bobcatfish** [comment]
Looks great @a-roberts , thanks so much!!

an awesome possum

**Comment by bobcatfish** [comment]
Oh no, @a-roberts it looks like you need to rebase again onto master :( I'm not sure why exactly this happens when there are no reported conflicts, but this seems to happen from time to time

**Comment by bobcatfish** [comment]
Looks like this PR is blocking Tide which doesn't seem to realize it can't merge - added a comment to https://github.com/kubernetes/test-infra/issues/10827 and for now:

/hold

We can cancel the hold as soon as the PR is rebased @a-roberts (It looks like you did rebase it already, or merged master in maybe? let me know if you need a 2nd pair of eyes on what's going on)

**Comment by a-roberts** [comment]
"Blocking tide"? How can my docs change do that? 

I merged master in yeah, feel free to work your magic!

**Comment by bobcatfish** [comment]
> "Blocking tide"? How can my docs change do that?

Sorry for the lack of context! Tide is one of the tools that make up Prow ([which we use for managing PRs](https://github.com/knative/build-pipeline/blob/master/CONTRIBUTING.md#pull-request-process)). Tide is specifically responsible for merging in PRs once they are in the right state (i.e. all checks pass and they have the approved + lgtm labels).

It has a bug (https://github.com/kubernetes/test-infra/issues/10827) around the way that we merge PRs - we merge by rebasing onto master, so unfortunately merging master into your branch isn't enough to fix the issue, you need to pull master and rebase onto it :S

Okay so I took a look at your PR and it looks like the problem is that some of the commits conflict with `master`. I did something like:

1. Updated my `master` branch from `upstream`
2. Checked out this branch
3. Did `git rebase master`

At that point I started running into conflicts, which I was trying to resolve but then I realized that this branch has 13 commits, so I want to suggest an easier way forward:

@a-roberts what do you think about creating a new branch, copying your changes to `tutorial.md` into the branch, and then opening a new PR with just one commit? The alternative is to try to rebase the whole branch onto master and deal with the conflicts one at a time, but I don't think it's worth it and in this case [having just one descriptive commit](https://github.com/knative/build-pipeline/blob/master/CONTRIBUTING.md#commit-messages) I think would be nicer in our history anyway.

(_Apologies cuz you probably are aware of this already but I also want to share that personally [I perfer rebasing onto master vs. merging master in](https://www.atlassian.com/git/tutorials/merging-vs-rebasing) b/c it generally leads to a cleaner project history imo and also makes it easier to move the commits around if you want to e.g. squashing._)

Sorry for all this complication @a-roberts , if you'd rather not make the new PR let me know and I can do it, but then you wouldn't be the author :(

**Comment by a-roberts** [comment]
@bobcatfish thanks for clarifying, done
